File: snippets/python/graph/dijkstra.py
from heapq import heappush, heappop

def dijkstra(
        graph: list[list[tuple[int, int]]] | dict[int, list[tuple[int, int]]],
        start: int
) -> list[int | float]:
    """
    Single source shortest path for non-negative weights
    graph[u] = [(v, weight), ...]
    Returns dist array from start
    Time: O((V + E) log V)
    """
    n = len(graph)
    dist = [float('inf')] * n
    dist[start] = 0
    pq = [(0, start)]
    
    while pq:
        d, u = heappop(pq)
        if d > dist[u]:
            continue
        for v, w in graph[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                heappush(pq, (dist[v], v)) # pyright: ignore
    return dist

File: snippets/rust/graph/dijkstra.rs
// Single-source shortest paths on a graph with non-negative weights.
// `g[u]` is a list of (v, w) outgoing edges. Returns vec of distances; INF = i64::MAX.
fn dijkstra(g: &Vec<Vec<(usize, i64)>>, s: usize) -> Vec<i64> {
    use std::cmp::Reverse;
    use std::collections::BinaryHeap;
    let n = g.len();
    let mut d = vec![i64::MAX; n];
    d[s] = 0;
    let mut pq: BinaryHeap<Reverse<(i64, usize)>> = BinaryHeap::new();
    pq.push(Reverse((0, s)));
    while let Some(Reverse((dist, u))) = pq.pop() {
        if dist > d[u] { continue; }
        for &(v, w) in &g[u] {
            let nd = dist + w;
            if nd < d[v] {
                d[v] = nd;
                pq.push(Reverse((nd, v)));
            }
        }
    }
    d
}
